File: androidApp/src/main/java/com/brettm/holiday/android/CountryInfoViewModel.kt
package com.brettm.holiday.android

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.brettm.holiday.holidayApi.HolidayApiFactory
import com.brettm.holiday.holidayApi.apis.CountryApi
import com.brettm.holiday.holidayApi.apis.PublicHolidayApi
import com.brettm.holiday.holidayApi.countryInfoUsingDeviceLocale
import com.brettm.holiday.holidayApi.nextPublicHolidaysUsingDeviceLocale
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.joinAll
import kotlinx.coroutines.launch
import org.openapitools.client.models.CountryInfoDto
import org.openapitools.client.models.PublicHolidayV3Dto

enum class CountryInfoContentKey {
    Info, Holidays
}

// Kotlin doesn't allow private/file private type aliases, which leads to verbose
// type aliases like this!
typealias CountryInfoViewModelContent = Map<CountryInfoContentKey, Any>

interface CountryInfoViewModelInterface: ViewModelInterface<CountryInfoViewModelContent> {
    val countryInfo: CountryInfoDto?
    val holidays: List<PublicHolidayV3Dto>?
}

class CountryInfoViewModel(
    private val countryClient: CountryApi = HolidayApiFactory.createCountryApi(),
    private val holidayClient: PublicHolidayApi = HolidayApiFactory.createPublicHolidayApi(),
    private val dispatcher: CoroutineDispatcher = Dispatchers.IO
) : ViewModel(), CountryInfoViewModelInterface {

    private val _uiState = MutableStateFlow<UIState<CountryInfoViewModelContent>>(UIState.Loading(null))
    override var uiState: StateFlow<UIState<CountryInfoViewModelContent>> = _uiState.asStateFlow()

    override val countryInfo: CountryInfoDto?
        get() {
            return uiState.value.content()?.get(CountryInfoContentKey.Info) as CountryInfoDto?
        }

    @Suppress("UNCHECKED_CAST")
    override val holidays: List<PublicHolidayV3Dto>?
        get() {
            return uiState.value.content()
                ?.get(CountryInfoContentKey.Holidays) as List<PublicHolidayV3Dto>?
        }

    override fun updateContent() {
        _uiState.value = UIState.Loading(uiState.value.content())
        viewModelScope.launch(dispatcher) {
            try {
                _uiState.value = UIState.Loaded(fetchContent())
            } catch (e: Exception) {
                _uiState.value = UIState.Error(e.toString())
            }
        }
    }

    @Throws(Exception::class)
    private suspend fun fetchContent(): CountryInfoViewModelContent {
        val content: MutableMap<CountryInfoContentKey, Any> = mutableMapOf()
        val jobs = listOf(
            viewModelScope.launch(dispatcher) {
                val countryInfoResult = fetchContent() {
                    countryClient.countryInfoUsingDeviceLocale()
                }
                content[CountryInfoContentKey.Info] = countryInfoResult.getOrThrow()
            },
            viewModelScope.launch(dispatcher) {
                val holidayResult = fetchContent() {
                    holidayClient.nextPublicHolidaysUsingDeviceLocale()
                }
                content[CountryInfoContentKey.Holidays] = holidayResult.getOrThrow()
            }
        )
        jobs.joinAll()
        return content
    }
}

File: iosApp/iosApp/CountryInfoViewModel.swift
import Foundation
import shared

enum CountryInfoContentKey: String {
    case info
    case holidays
}

@MainActor
protocol CountryViewModelInterface: ViewModelInterface where Content == [CountryInfoContentKey : Any] {
    var countryInfo: CountryInfoDto? { get }
    var holidays: [PublicHolidayV3Dto]? { get }
}

@Observable
@MainActor
class CountryInfoViewModel: CountryViewModelInterface, ContentFetching {

    typealias Content = [CountryInfoContentKey : Any]
    typealias Response = Any
    
    public var uiState: UIState<Content> = UIState.loading(nil)
        
    public var countryInfo: CountryInfoDto? {
        return uiState.content()?[.info] as! CountryInfoDto?
    }
    public var holidays: [PublicHolidayV3Dto]? {
        return uiState.content()?[.holidays] as! [PublicHolidayV3Dto]?
    }
    
    private let countryClient = HolidayApiFactory.companion.createCountryApi()
    private let holidayClient = HolidayApiFactory.companion.createPublicHolidayApi()
    
    nonisolated func fetchContent() async throws -> Content {
        return try await withThrowingTaskGroup(of: Content.self) { group in

            group.addTask{
                let countryInfo = try await self.fetchResponse {
                    try await self.countryClient.countryInfoUsingDeviceLocale()
                }
                return [CountryInfoContentKey.info: countryInfo]
            }
            group.addTask{
                let holidays = try await self.fetchResponse {
                    try await self.holidayClient.nextPublicHolidaysUsingDeviceLocale()
                }
                return [CountryInfoContentKey.holidays: holidays]
            }
        
            var content: [CountryInfoContentKey : Any] = [:]
            for try await value in group {
                content.merge(value, uniquingKeysWith: {_, new in new})
            }
            return content
        }
    }
}
